All_materials/20230822_Web_Scraping_Financial_Ombudsman_Service/src/web_scraping_and_data_wrangling/fos_web_scrap.py

```diff
@@ -0,0 +1,335 @@
+        
+        #Getting the info
+        ref_number = each_deison[0][19:]
+        date_of_desision = datetime.datetime.strptime(each_deison[1], '%d %b %Y')
+        bank = each_deison[2]
+        outcome = each_deison[5].replace(" ", "")
+        descrp = each_deison[8].split(' ', 2)[2][:190]
+        pdf_text = getting_pdf_text(ref_number)
+        
+        dict_each_deision = {
+            'ref':ref_number,
+            'date':date_of_desision,
+            'scale_of_lost': finding_scale_of_money(pdf_text),
+            'bank':bank,
+            'outcome':outcome,
+            'background':descrp,
+            'all_text': pdf_text,
+            'CRM':was_CRM_mentioned(pdf_text)
+            }
+        
+        list_of_machine_readable_deision = list_of_machine_readable_deision + [dict_each_deision]
+        
+    return list_of_machine_readable_deision
+
+def getting_number_of_links_and_soup(soup, url_link):
+    """
+    Getting the list from soup
+
+    Parameters
+    ----------
+    soup 
+    
+    url
+
+    Returns
+    -------
+    number_of_pages
+        Number of pages
+    list_of_urls
+        The list the URLs
+
+    """
+    
+    #Getting the number of pages
+    out = soup.find_all("div", {"class": "search-results-holder"})[0]
+    number_of_pages = int(out.text.split("\n")[-5].split()[-1])
+    
+    #Getting the URL
+    list_of_urls = [url_link + '&Start=' + str(x*10) for x in range(1,number_of_pages)]
+    list_of_urls = [url_link] + list_of_urls
+    
+    return list_of_urls, number_of_pages
+
+def get_complete_desision_list(url_link):
+    
+    #Getting all the 0th soup
+    soup = get_info_as_soup(url_link)
+    
+    links_for_all_lists, _ = getting_number_of_links_and_soup(soup, url_link)
+    
+    list_of_deisions = []
+    
+    for each_url in links_for_all_lists:
+        
+        each_soup = get_info_as_soup(each_url)
+        
+        desison_list = get_list_of_deision(each_soup)
+
+        list_of_machine_readable_deision = wrangle_each_deision_list(desison_list)
+        
+        list_of_deisions = list_of_deisions + list_of_machine_readable_deision
+        
+    return list_of_deisions
+
+def getting_pdf_text(ref_complaint):
+    
+    #Getting the pdf link
+    url_pdf = 'https://www.financial-ombudsman.org.uk/decision/' + ref_complaint + '.pdf'
+    
+    #request the download of the pdf and getting the text
+    response = requests.get(url_pdf)
+    my_raw_data = response.content
+    
+    all_text = ''
+    
+    with BytesIO(my_raw_data) as data:
+        read_pdf = PyPDF2.PdfReader(data)
+
+        for page in range(len(read_pdf.pages)):
+            each_page_text = read_pdf.pages[page].extract_text()
+            all_text = all_text + each_page_text
+            
+    return all_text
+
+def getting_numbers_from_string(string_with_numbers):
+    
+    numbers_only = [x for x in string_with_numbers if x.isdigit()]
+    numbers_only = ''.join(numbers_only)
+    
+    return numbers_only
+
```

Comment:
all functions should have double blank lines between them

Reply:
fixed